internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -34,55 +37,71 @@ func (b *Builder) BuildDestinations(destinations []*intermediate.Destination) *B
+
+	return b.addEndpointsRef(sni, destination.ServiceEndpoints.Resource.Id, portInfo.meshPortName).
+		addRouters(portInfo, destination, serviceRef, sni, b.localDatacenter, true).
+		addClusters(portInfo, destination, sni)
+}
+
+// addClusters creates clusters for each service port in the pre-processed a servicePortInfo.
+func (b *Builder) addClusters(portInfo *servicePortInfo, destination *intermediate.Destination, sni string) *Builder {
+	for portName, port := range portInfo.servicePorts {
+		if port.GetProtocol() != pbcatalog.Protocol_PROTOCOL_TCP {
+			//only implementing L4 at the moment
+		} else {
+			clusterName := fmt.Sprintf("%s.%s", portName, sni)
+			b.addCluster(clusterName, sni, portName, destination.Identities)
 		}
 	}
+	return b
 }
 
-func (b *Builder) buildImplicitDestination(destination *intermediate.Destination) {
-	serviceRef := resource.Reference(destination.ServiceEndpoints.Resource.Owner, "")
-	clusterName := DestinationClusterName(serviceRef, b.localDatacenter, b.trustDomain)
-	statPrefix := DestinationStatPrefix(serviceRef, b.localDatacenter)
-
-	// We assume that all endpoints have the same port protocol and name, and so it's sufficient
-	// to check ports just from the first endpoint.
-	if len(destination.ServiceEndpoints.Endpoints.Endpoints) > 0 {
-		// Find the destination proxy's port.
-		// Endpoints refs will need to route to mesh port instead of the destination port as that
-		// is the port of the destination's proxy.
-		meshPortName := findMeshPort(destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports)
-
-		for _, port := range destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports {
-			b.outboundListenerBuilder.
-				addRouterWithIPMatch(clusterName, statPrefix, port.Protocol, destination.VirtualIPs).
-				buildListener().
-				addCluster(clusterName, destination.Identities).
-				addEndpointsRef(clusterName, destination.ServiceEndpoints.Resource.Id, meshPortName)
+// addRouters creates routers for each service port in the pre-processed a servicePortInfo.
+func (b *Builder) addRouters(portInfo *servicePortInfo, destination *intermediate.Destination,
+	serviceRef *pbresource.Reference, sni, datacenter string, isImplicitDestination bool) *Builder {
+
+	for portName, port := range portInfo.servicePorts {
+		statPrefix := DestinationStatPrefix(serviceRef, portName, datacenter)
+
+		if port.GetProtocol() != pbcatalog.Protocol_PROTOCOL_TCP {
+			//only implementing L4 at the moment
+			continue
 		}
+
+		clusterName := fmt.Sprintf("%s.%s", portName, sni)
+		var portForRouterMatch *pbcatalog.WorkloadPort
+		// router matches based on destination ports should only occur on implicit destinations
+		// for explicit, nil will get passed to addRouterWithIPAndPortMatch() which will then
+		// exclude the destinationPort match on the listener router.
+		if isImplicitDestination {
+			portForRouterMatch = port
+		}
+		b.addRouterWithIPAndPortMatch(clusterName, statPrefix, portForRouterMatch, destination.VirtualIPs)
```

Comment:
For tproxy's outbound listener, this port should be the virtual port from the service instead. We can address it in a follow-up PR though

Reply:
Added NET-5587 to track this.